demo/src/pick_place_task.cpp

```diff
@@ -0,0 +1,470 @@
+	 *               Load Parameters                    *
+	 *                                                  *
+	 ***************************************************/
+	ros::NodeHandle pnh("~");
+
+	// Planning group properties
+	arm_group_name_ = pnh.param<std::string>("arm_group_name", "arm_group_name");
+	hand_name_ = pnh.param<std::string>("hand_name", "hand_group_name");
+	eef_name_ = pnh.param<std::string>("eef_name", "eef_name");
+	hand_frame_ = pnh.param<std::string>("hand_frame", "hand_frame_name");
+	world_frame_ = pnh.param<std::string>("world_frame", "world_frame_name");
+
+	// poses
+	open_gripper_pose_ = pnh.param<std::string>("open_gripper_pose", "open_pose");
+	close_gripper_pose_ = pnh.param<std::string>("close_gripper_pose", "close_pose");
+	home_pose_ = pnh.param<std::string>("home_pose", "home_pose");
+
+	// Object + surface
+	table_name_ = pnh.param<std::string>("table_name", "table_name");
+	table_height_ = pnh.param<double>("table_height", 0.0);
+	table_length_ = pnh.param<double>("table_length", 0.0);
+	table_width_ = pnh.param<double>("table_width", 0.0);
+
+	object_name_ = pnh.param<std::string>("object_name", "name_of_moved_object");
+	object_height_ = pnh.param<double>("object_height", 0.0);
+	object_radius_ = pnh.param<double>("object_radius", 0.0);
+
+	table_refrence_frame_ = pnh.param<std::string>("table_refrence_frame", "frame_table_is_in");
+	object_refrence_frame_ = pnh.param<std::string>("object_refrence_frame", "frame_object_lays_upon");
+	surface_link_ = pnh.param<std::string>("surface_link", "name_of_table");
+	support_surfaces_ = { surface_link_ };
+
+	// Pick
+	approach_object_min_dist_ = pnh.param<double>("approach_object_min_dist", 0.0);
+	approach_object_max_dist_ = pnh.param<double>("approach_object_max_dist", 0.0);
+
+	// Lift
+	lift_object_min_dist_ = pnh.param<double>("lift_object_min_dist", 0.0);
+	lift_object_max_dist_ = pnh.param<double>("lift_object_max_dist", 0.0);
+
+	// Place
+	place_surface_offset_ = pnh.param<double>("place_surface_offset", 0.0);
+	place_pos_x_ = pnh.param<double>("place_pos_x", 0.0);
+	place_pos_y_ = pnh.param<double>("place_pos_y", 0.0);
+
+	// compute hand grasp frame
+	double rotationy = pnh.param<double>("grasp_rotation_y", 0.0);
+	double rotationx = pnh.param<double>("grasp_rotation_x", 0.0);
+	double grasp_offset_x = pnh.param<double>("grasp_offset_x", 0.0);
+	double grasp_offset_y = pnh.param<double>("grasp_offset_y", 0.0);
+	double grasp_offset_z = pnh.param<double>("prasp_offset_z", 0.0);
+	grasp_frame_transform_ = Eigen::AngleAxisd(M_PI * rotationy, Eigen::Vector3d::UnitY()) *
+	                         Eigen::AngleAxisd(M_PI * rotationx, Eigen::Vector3d::UnitX()) *
+	                         Eigen::Translation3d(grasp_offset_x, grasp_offset_y, grasp_offset_z);
+}
+
+void PickPlaceTask::init() {
+	const std::string object = "object";
+
+	// Reset ROS introspection before constructing the new object
+	// TODO(henningkayser): verify this is a bug, fix if possible
```

Comment:
This is a known problem and we hit it before.
A patch is very welcome, if you have a good idea on how to improve this without too much overhead.

Reply:
We talked about this before @rhaschke .
If you reset a shared-pointer to a task with a `new Task` you create in the same line,
the new Task (and its introspection) are setup before the destructor of the old one got called.
This becomes a problem as only one ROS service can be active with the same name in the same node and the new one just does not get set up.

Reply:
Thanks for this clarification. One option to solve the issue in MTC is to use a global cache of those ROS services (and reuse existing ones).
File an issue: https://github.com/ros-planning/moveit_task_constructor/issues/106.

Reply:
Yes, but then we add static state to the Task and I believe we have avoided it thus far.
Not counting your planner cache which I still plan to remove again.

Reply:
I don't consider a cache as static state. Formally, of course, every cache is.
Late initialization will be brittle on timing. Lazy might work, but requires much more work.
Currently, the existence of the introspection pointer also indicates that introspection is desired...